app/domain/authentication/authn_gce/update_authenticator_input.rb

```diff
@@ -0,0 +1,87 @@
+require 'command_class'
+
+module Authentication
+  module AuthnGce
+
+    UpdateAuthenticatorInput = CommandClass.new(
+      dependencies: {
+        verify_and_decode_token: Authentication::OAuth::VerifyAndDecodeToken.new,
+        decoded_token_class:     DecodedToken,
+        logger:                  Rails.logger
+      },
+      inputs:       [:authenticator_input]
+    ) do
+
+      extend Forwardable
+      def_delegators :@authenticator_input, :authenticator_name, :account, :credentials, :username
+
+      def call
+        validate_token
+        updated_input
+      end
+
+      private
+
+      def validate_token
+        decoded_token
+        validate_audience
+      end
+
+      def decoded_token
+        @decoded_token ||= @decoded_token_class.new(
+          decoded_token_hash: @verify_and_decode_token.call(
+            provider_uri:     PROVIDER_URI,
+            token_jwt:        decoded_credentials.jwt,
+            claims_to_verify: {
+              verify_iss:        true,
+              iss:               PROVIDER_URI,
+              verify_iat:        true,
+              verify_expiration: true
+            }
+          ),
+          logger:             @logger
+        )
+      end
+
+      def decoded_credentials
+        @decoded_credentials ||= Authentication::Jwt::DecodedCredentials.new(credentials)
+      end
+
+      def validate_audience
+        if audience_parts.length != 3 ||
+            audience_parts[0] != "conjur" ||
+            audience_parts[1] != account
+          raise Errors::Authentication::AuthnGce::InvalidAudience, audience
+        end
+      end
+
+      def audience_parts
+        @audience_parts ||= audience.split('/', 3)
+      end
+
+      def audience
+        @audience ||= @decoded_token.audience
+      end
+
+      def conjur_username
+        return @conjur_username if @conjur_username
+
+        @conjur_username = audience_parts[2]
+
+        @logger.debug(
+          LogMessages::Authentication::Jwt::ExtractedUsernameFromToken.new(
+            conjur_username
+          )
+        )
+        @conjur_username
+      end
+
+      def updated_input
```

Comment:
`updated_input` a couple of notes here
1. purposeful past tense?
2. doesn't explain what we are updating and for what. Why are we updating the authenticator object with items from the token? @orenbm Is this like xms_mirid for authn-azure to identify the machine?
3. I would add a sentence or two at the top of CommandClass explaining why this update is necessary

Reply:
1. it's not past tense. it's a noun as this method returns an updated input.
2. it is more like `authn-oidc`. the reason for this is that in the general `Authenticate` command class we perform validations on the username, but the username is retrieved from the token. So we first decode the token and retrieve the username and update the authenticator input with it. The credentials is also updated to have the decoded token instead of the original so we don't need to decode it again.
3. will do, thanks!

Reply:
1. What about `update_input_from_token`?
2. great! 
3. great! x2

Reply:
1. the method returns an object so its name needs to be a noun. if it was performing an action then the name would be an action.